numpyro/distributions/flows.py

```diff
@@ -0,0 +1,25 @@
+class BlockNeuralAutoregressiveTransform(Transform):
+    event_dim = 1
+
+    def __init__(self, bna_nn, params):
+        self.bna_nn = bna_nn
+        self.params = params
+
+    def __call__(self, x):
+        self._cached_x = x
+        y, self._cached_logdet = self.bna_nn(self.params, x)
```

Comment:
@neerajprad I would like to mimic your solution at #239 to implement `call_with_intermediates` which returns both `y` and `log_det`; and `log_abs_det_jacobian(self, x, y, intermediates=None)` which returns logdet from `intermediates`. What do you think about it? Is there a better solution which you can think about?

Reply:
There may be a better pattern, but this seems like a reasonable first approach. I am not sure if `log_abs_det_jacobian` needs to be modified since the callee can probably just store and reuse the `log_det` terms. We can also modify `TransformedDistribution.sample_with_intermediates` to use `call_with_intermediates` and pass the list of (x, log_det) terms to `log_prob` as intermediates.